[Quant] Add DQ duplication pass

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107107
* __->__ #107900
* #107899
* #107106
* #107105

Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

kimishpatel committed:
[Quant] Add DQ duplication pass

Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

jerryzh168 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -11,12 +11,83 @@
+    return False
+
+
+def _find_q_dq_node_for_user(
```
I feel this will be hard to debug and maintain, what is the problem with the existing duplicate dq implementation: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/qat_utils.py#L580?

jerryzh168 replied:
oh I looked again, is this for test only? can you move this to test files?

kimishpatel replied:
No this is also used later in the stack for metdata porting.

jerryzh168 commented on torch/ao/quantization/pt2e/duplicate_dq_pass.py:
```diff
@@ -0,0 +1,76 @@
+    _copy_and_replace_use(gm, producer, user)
+
+
+class DuplicateDQPass(_ExportPassBase):
```
I'm wondering if it makes sense to just always duplicate dequant for now, or do a duplicate and then deduplicate like we have did before in https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/qat_utils.py#L580-L627

kimishpatel replied:
Oh you mean, move this before fold_conv_bn? If so yes. makes sense.

jerryzh168 replied:
I was wondering if we can do:
```
duplicate all dequant -> port metadata -> deduplicate dequantize ops when we didn't find metadatas on them
```

kimishpatel replied:
why is that better? It might be cleaner if we dont duplicate and dedup.

jerryzh168 replied:
so current quantization_annotation doesn't really convey any information about share or not share dq nodes, I think it will be better to not rely on it

actually I'm wondering when does this have to be landed? I'm thinking of doing a proper design for this part next week, right now this information is missing from QuantizationSpec

kimishpatel replied:
there is no sharing of dq nodes at all. The only reason annotation is taken into account is so that you dont unnecessarily duplicate dq nodes. I can very much remove quantization_annotation from decision making, it will just have performance imlication

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -11,12 +11,83 @@
     "fold_bn_weights_into_conv_node",
     "get_aten_graph_module",
     "remove_tensor_overload_for_qdq_ops",
+    "filter_sym_size_users",
+    "find_q_dq_node_for_user",
+    "is_valid_annotation",
```
nit: could you make these private?

kimishpatel replied:
I did. and having to export it from util was complaining that if this is to be used elsewhere remove `_`. If you have suggestions let me know and i can do it in subsequent diffs

jerryzh168 replied:
I see, maybe I can try to remove, I do remember we have been successfully using private utils somewhere

jerryzh168 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -11,12 +11,83 @@
+    return False
+
+
+def find_q_dq_node_for_user(
```
I feel this function is a bit too complicated, is there a way to simplify this?

kimishpatel replied:
it is just finding q, dq from [producer -> .... -> q -> dq - > user]. I can add doc string if you want

jerryzh168 replied:
yeah a docstring would be helpful

jerryzh168 commented on torch/ao/quantization/pt2e/duplicate_dq_pass.py:
```diff
@@ -0,0 +1,73 @@
+"""
+
+
+def _copy_and_replace_use(
```
also maybe merge this function with `_maybe_duplicate_dq`?

kimishpatel replied:
Sure

jerryzh168 commented on torch/ao/quantization/pt2e/duplicate_dq_pass.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def _maybe_duplicate_dq(
+    gm: torch.fx.GraphModule, producer: torch.fx.Node, user: torch.fx.Node
```
nit: renaming `producer` to `dq_node` might be clearer

kimishpatel replied:
sounds good

jerryzh168 commented on torch/ao/quantization/pt2e/duplicate_dq_pass.py:
```diff
@@ -0,0 +1,73 @@
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        for node in graph_module.graph.nodes:
+            if node.op == "call_function" and node.target in _DEQUANTIZE_OPS:
+                dq_users = filter_sym_size_users(node)
```
is this needed? can we just replace all users? also after the IR changes, we don't have sym size nodes anymore right?

kimishpatel replied:
well I dont know if the sym size node will show up due to user code. like querying size on a tensor. I think here we are just making a call of not duplicating dq for sym size nodes. Either choice is fine. I would say we can go with this for now and change the behavior later

jerryzh168 commented on torch/ao/quantization/pt2e/duplicate_dq_pass.py:
```diff
@@ -0,0 +1,73 @@
+    """
+
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        for node in graph_module.graph.nodes:
```
also I'm wondering if you can reuse https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/qat_utils.py#L585 actually, I think we don't need to maintain multiple implementations of the same thing

kimishpatel replied:
So that doesnt account for all the things that is being done here. things like sym_size filtering, accounting valid annotation etc. I agree with unifying the code though, but I would imagine we can update that code or move the fold_conv_bn after duplicate dq. although that needs some code update because in fold conv bn it removes duplicate dq at some point.

kimishpatel commented on torch/ao/quantization/pt2e/duplicate_dq_pass.py:
```diff
@@ -0,0 +1,73 @@
+
+
+"""
+This API is specifically to replace a specific use of producer node
```
yes

mcr229 commented:
Does this mean DQ duplication will be a default in prepare pt2e?

jerryzh168 commented:
> Does this mean DQ duplication will be a default in prepare pt2e?

yeah, this will be handled in quant flow for normal cases

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/__init__.py:
```diff
@@ -12,7 +12,6 @@
 from .quantize import *  # noqa: F403
 from .quantize_jit import *  # noqa: F403
 from .stubs import *  # noqa: F403
-from .pt2e.utils import move_model_to_eval
```
@andrewor14 i had to make this change otherwise, I was running into circular import due to 
`from torch.ao.quantization.quantizer import QuantizationAnnotation` inside pt2e.utils.py

andrewor14 replied:
Wait then this is a BC breaking change since we're landing this in 2.1. I think we should find a way to work around this

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

andrewor14 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -10,13 +10,98 @@
+    return node_users
+
+
+def _is_valid_annotation(annotation: QuantizationAnnotation) -> bool:
```
This is the only place where we use `QuantizationAnnotation` right? Maybe we can leave this as `Any` for now since it's an internal function?

kimishpatel replied:
Let me see if there is a work around

andrewor14 replied:
Hey @kimishpatel, I have this PR which should take care of the circular deps issue: https://github.com/pytorch/pytorch/pull/108363. Please rebase on top of this when you have the chance, then you don't have to do anything special

andrewor14 reviewed (changes requested):
Sorry Kimish but I think we want to keep `move_model_to_eval` under `torch.ao.quantization` directly (cc @jerryzh168). We're also cherry picking this back in 2.1 so it would break BC for those users if we moved this. Could we look into other ways to work around the circular dependency issue?

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

kimishpatel commented:
@leslie-fang-intel can you help look at this failure https://github.com/pytorch/pytorch/actions/runs/6043973442/job/16402138568?pr=107900?

leslie-fang-intel commented:
> 

Sure, I will take a look soon @kimishpatel

leslie-fang-intel commented:
Hi @kimishpatel, I think we need to make corresponding changes in `test_mkldnn_pattern_matcher.py`. Please kindly refer to this commit: https://github.com/leslie-fang-intel/pytorch/commit/59e5ce92f8b5ca68205df0abc444d0b55f16fc58 to adjust the match numbers. Please let me if anything missing.

andrewor14 reviewed (approved):
`move_model_to_eval` changes look good to me

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

kimishpatel commented:
> Hi @kimishpatel, I think we need to make corresponding changes in `test_mkldnn_pattern_matcher.py`. Please kindly refer to this commit: [leslie-fang-intel@59e5ce9](https://github.com/leslie-fang-intel/pytorch/commit/59e5ce92f8b5ca68205df0abc444d0b55f16fc58) to adjust the match numbers. Please let me if anything missing.

Thanks Leslie. Please review changes to `test_mkldnn_pattern_matcher.py`.

As I was reviewing those, I realize that decomposition for quantize/dequantize does not match that of here, https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/representation/rewrite.py#L428. Specifically in inductor decomp of quantize_per_tensor, addition of zero point still happens in fp, which I dont think is accurate because of the rounding errors that can happen on rouding zero point to fp and then whatever rounding happens during add. So I think we should sync those. cc: @jerryzh168 

Also, please see how this diff affects quantization for x86 inductor. I see there was a pass to remove convert node that appear in quant/dequant pattern. But since now quant can more than 1 dequant users, that pattern wont match. If you still need to remove such redundant convert nodes, you will have to change the pattern/replacement

kimishpatel committed:
Update on "[Quant] Add DQ duplication pass"


Summary:
During convert step observers are first replaced by Q-DQ pair. In some
scenarios like following output DQ has a fan out.

                 ---> OP2 -> Q -> DQ
                /
OP -> Q -> DQ -
                \
                 ---> OP3 -> Q -> DQ

If either op OP2 or OP3 are configured to be quantized, then the input
is expected to quantized. In this case quantized equivalent of some
pattern, that quantizer asked to be quantized, should look like:
[DQ -> {pattern} -> Q]. However, in scenario like above where DQ node
is shared between multiple "quantized" patterns, boundary of "quantized"
pattern is not clear because DQ now belongs to multiple quantized
patterns.

This poses challenge for:
- Porting metadata: which "quantized" partition this DQ node belongs
- Quantized representation, equivalently, needs to identify
self-contained quantized pattern that is replaced by its equivalent pattern
that captures compute in the quantized precision.

Test Plan:
test_duplicate_dq_pass

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48663147](https://our.internmc.facebook.com/intern/diff/D48663147)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48663147).

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -513,6 +513,13 @@ def forward(self, x):
             # 1. Pair of to_int8 and to_fp32 at conv input * 1, extra input of add * 1, and graph output * 1
             #    matched in pointless_convert pass at
             #    torch/_inductor/fx_passes/joint_graph.py: [convert_element_type, convert_element_type_1]
+            #    NB: since quant workflow now duplicates DQ node, for each user, we wont necessarily see
```
Hi @kimishpatel, could you also update the comments here https://github.com/leslie-fang-intel/pytorch/blob/59e5ce92f8b5ca68205df0abc444d0b55f16fc58/test/inductor/test_mkldnn_pattern_matcher.py#L514-L515 and other similar places? so, it will be aligned with expected count values.
